ledger/src/shred.rs

```diff
@@ -1296,6 +1299,28 @@ pub fn should_discard_shred(
     false
 }
 
+// Drains coding shreds out of the vector of shreds.
+// Note that the function does not preserve the order of either the retained or
+// the drained shreds.
+// TODO: Use Vec::extract_if instead once stable.
+pub(crate) fn drain_coding_shreds(shreds: &mut Vec<Shred>) -> Drain<'_, Shred> {
+    let (mut i, mut j) = (0, shreds.len().saturating_sub(1));
+    loop {
+        while i < j && shreds[i].is_data() {
+            i += 1;
+        }
+        while i < j && shreds[j].is_code() {
+            j -= 1;
+        }
+        if i < j {
```

Comment:
Checking my understanding, in practice this conditional should never be executed since `merkle::recover` will return all data shreds before coding shreds?
This just covers our bases in case this fn is used elsewhere

Reply:
right, I also didn't want `blockstore::handle_shred_recovery` rely on that presumption because we can't guarantee that it won't change.

e.g. if we `flatten` and `collect::<Vec<_>>` earlier in the pipeline, then`Vec<Shred>` here is concatenation of multiple erasure batches and coding shreds can be in between data shreds.